// youban-utils/settings/setting.js
module.exports = {
  // 正式环境
  title: "有伴科技", // 标题
  defaultAllow: "?", // 缺省所有的页面必须登录
  urls: {
    // 地址配置
    login: null // 统一登录地址
  },
  servers: {
    // 服务器地址列表
    commonServer: "/common-api", //基础服务
    userServer: "/user-api", //用户服务
    statServer: "/stat-api", //统计服务
    oaServer: "/oa-api", //办公服务
    attendanceServer: "/attendance-api", //考勤排班服务
    medicalServer: "/medical-api", //医护服务
    fireServer: "/fire-api", //消防服务
    elderServer: "/elder-api", //老人服务
    mediaServer: "/media-api", //媒体服务，比如上传文件等
    financeServer: "/finance-api" //结算服务
  },
  getServer: function(serverName) {
    if (!serverName) {
      console.error("getServer() required serverName!");
      return "";
    }
    let url = serverName.toLowerCase();
    if (
      url.startsWith("http://") ||
      url.startsWith("https://") ||
      url.startsWith("//")
    ) {
      return serverName;
    }
    url = this.servers[serverName];
    if (!url) {
      console.error("getServer() server [" + serverName + "] url not defined!");
    }
    return url;
  }
};
